### ISSUE TITLE
Controller is always disconnecting randomly when using Bluetooth.

### ISSUE THREAD

**Issue by Newtype14** [Ryochan7/DS4Windows#1420]
Log: [DS4Windows Error 1167.txt](https://github.com/Ryochan7/DS4Windows/files/4980074/DS4Windows.Error.1167.txt)

I get this Error 1167 every time I play using Bluetooth. I'm forced to always play through USB connection because of this. The error doesn't occur if I turn off `Enable Output data to DS4`, so I know it's not a problem with my Bluetooth connection. I had this problem with version 2.1.3 as well, but I didn't try other versions before that.

**Comment by mika-n** [comment]
Maybe the debug version tells more info in a generated debug logfile?
https://github.com/Ryochan7/DS4Windows/wiki/Troubleshooting#detailed-debug-log-file-from-gamepad-connection-problems

Anyway. Based on the logfile you posted already tells that you have DS4 revision1 gamepad. Is this new gamepad or 5-6 years old DS4 gamepad? The thing is that Sony haven't manufactured those DS4 rev1 gamepads in several years after they released "fixed" DS4 revision2 gamepad.  

Do you have genuine Sony gamepad? During the last couple years there has been many fake Chinese wanna-be DS4 gamepad copies using the official Sony DS4 revision1 hardware ID. These gamepads often have issues over BT connection (works just fine over USB, but BT is unreliable because of poor "BT radio"). 

What ever it is then the debug version may tell a bit more why the connection is unreliable. This same issue has been discussed several times in other issue posts and there is not much developers can do without actually having one of those fake copies on hand while debugging the issue. There might be way to make those fake copies work better, but someone with such a device should debug it.

**Comment by Newtype14** [comment]
I verified my controller and it's authentic. I have another DS4, the one that came with my PS4, but it's also Rev 1. They're both controllers that I bought before Rev 2 came out, but they're not the Rev 1s from release day that had inferior joysticks and stuff like that. They're "second generation" Rev 1. Unfortunately, I don't own a Rev 2 to do some testing.

I've been trying to use the debug version, but when I launch it, there's an UNKNOWN ID that is already there and in Windows it shows up as another XBOX 360 controller (no, it's not my other controller). Even before I connect my controller, there's already a ton of issues being detected in the log window because of that UNKNOWN ID. It's also preventing my controller from getting detected in RetroArch, probably 'cause RetroArch detects two XBOX 360 controllers at launch. Steam games seem to be running fine though.

**Comment by mika-n** [comment]
Logfile from the debug test run?  %APPDATA%\ds4windows\logs\

**Comment by Tsuzuriko** [comment]
I've been getting the same error, but I'm pretty sure my DS4 is rev2, so it might not be related. 

A couple of things I can say that might help on pinpointing the issue: 

To get what I think is the most important bit of info out of the way, I've had this exact same problem with my old DS3 controller when using the updated SCPToolkit by nefarius (not the original ScarletCrush version). It just so happens that ViGEmBus which is used by Ryochan7's fork of DS4Windows is also partially made by nefarius. Whatever is causing this issue could be related to it, because at first I was using the outdated version of DS4Windows by Jays2Kings and my controller didn't disconnect once. I played over 1 hour with no issues whatsoever.

I'd be still using the original fork of DS4Windows, but when I proceeded to configure profiles for some of my emulators I noticed motion controls for Dolphin weren't supported, so I looked for an updated version and found this fork, which supports it. I proceeded to configure everything and it all worked as intended, but after a few minutes playing my controller disconnected mid-play. I turned it on once more and it reconnected, but the same thing happened again after another couple of minutes. Then I tried closing the emulator and playing the same game I was playing before updating to see if it helped, but ended up getting the same result shortly afterwards.

[DS4Windows Log (Error 1167).txt](https://github.com/Ryochan7/DS4Windows/files/4997610/DS4Windows.Log.Error.1167.txt)

**Comment by mika-n** [comment]
`07/29/2020 14:05:21: Gamepad data write connection is lost. Disconnecting the gamepad. LastErrorCode=1167
`
Looks like write-out fails for some reason. If you UNTICK "Enable output data to DS4" option in "other" tab page in a profile settings then does it work without disconnection issues?  However, unticking this option means that ligthbar and rumble doesn't work (ie. PC doesn't send out/write out data to DS4 gamepad). Anyway, there has been cases where simultaneous reads and writes with DS4 gamepad have caused issues in BT.

**Comment by Tsuzuriko** [comment]
> 
> 
> `Looks like write-out fails for some reason. If you UNTICK "Enable output data to DS4" option in "other" tab page in a profile settings then does it work without disconnection issues? `

I did as suggested and the controller disconnected again, but this time I got error 997 instead of 1167. I went ahead and downloaded the debug version to test this time, I don't know if it'll be any help but here's the more detailed log nonetheless.

[DS4Windows_Debug_Log_(Error 997).txt](https://github.com/Ryochan7/DS4Windows/files/5001900/DS4Windows_Debug_Log_.Error.997.txt)

**Comment by mika-n** [comment]
Do you still have those old ScpToolkit/ScarletCrush drivers installed? Those are known to cause problems in BT connections when some other "driver" tries to use the same BT host with gamepads. ScpTookit and ScarletCrush drivers hijack the BT stack and does all kind of nasty tricks (not recommended by Windows kernel driver docs).

Uninstall and traces of those old Scp drivers and apps (ie. literally remove those drivers using DriveStore Explorer app even when you no longer have apps using those drivers).
https://forums.vigem.org/topic/250/scptoolkit-removal-guide

**Comment by Tsuzuriko** [comment]
Sorry for the late reply, had some things to take care of irl and it slipped my mind to update this thread. I had already done the steps you suggested before I installed Jays2Kings DS4Windows version, but decided to do it again just in case. Turns out there was still residue from the installation, even though it didn't show up anywhere on my system. Though the steps suggested at first didn't yield any results, DevManView (which is suggested at the end of the guide to be used as a means of verification) not only detected the leftover registries, but also seems to have been able to get rid of them, even though nefarius doesn't suggest using it for that purpose. Repeating the steps on the guide as he suggests proved ineffective to me, so I gave DevManView a shot at removing them and it seems to have worked. 

After a few days of occasionally playing I haven't had disconnects again, only very rare latency spikes (once every few hours or so, lasting on average 2 seconds), but I'm fine with those. They might not even be caused by DS4Windows, could be my Bluetooth adapter or even Windows itself.

Thanks for the help, hopefully someone with a similar issue will be able to solve it in the future with the information in this thread.

**Comment by Newtype14** [comment]
Sorry for the very late reply. Got really busy all of a sudden and since this bug happens randomly, it requires a bit of time to test each time.

I also had some traces of SCPToolkit on my PC. I removed it all using the guide. At first I thought for sure that the solution had been found because I played for a long time without any disconnection. However, my controller got disconnected again while it was just resting on my desk (it wasn't idle disconnect). I wasn't even playing a game when it happened.

I also just found what was causing the UNKNOWN ID to appear in DS4Windows' Debug version. It's my Blue Yeti microphone. As you can see in the following log, HID#17, referred to as HID-compliant headset, is actually my Blue Yeti.

[DS4Windows Debug Log 01.txt](https://github.com/Ryochan7/DS4Windows/files/5049081/DS4Windows.Debug.Log.01.txt)

I unplugged my microphone and this is the log I get now when I launch DS4Windows:

[DS4Windows Debug Log 02.txt](https://github.com/Ryochan7/DS4Windows/files/5049083/DS4Windows.Debug.Log.02.txt)

I'm gonna be testing my controller again without the microphone to see if I'll get disconnects. You never know. In the meantime, is there anything I can do to prevent DS4Windows from detecting my microphone in the future?

**Comment by mika-n** [comment]
@Newtype14   The BlueYeti device is shown as "gamepad" only in debug version because the debug version tries to use all devices with usageFlag=0x05 value. The vanilla version doesn't try to use it as gamepad because the VID/PID is not registered as DS4 gamepad.

**Comment by Newtype14** [comment]
Unfortunately, I'm still getting disconnects. Here's the log: [DS4Windows Debug Log.txt](https://github.com/Ryochan7/DS4Windows/files/5054865/DS4Windows.Debug.Log.txt)

I'm gonna try to get my controller to disconnect without using DS4Windows. If it disconnects, then I'll know that DS4Windows is not the problem.

**Comment by mika-n** [comment]
The difference between running with DS4Windows and without DS4Windows app is that "standard" WinOS DS4 gamepad driver doesn't try to use lightbar or rumble effects over BT connection. Therefore WinOS doesn't try to do "writeOutput to DS4 gamepad" things at all while connected over BT (you would need the special Sony WSA wireless dongle to do that without DS4Windows app). Using a standard BT dongle/chipset means that PC only reads data from DS4 and does not try to write out rumble/lightbar data over BT UNLESS you use DS4Windows app.

So, could be that you won't see any disconnections without DS4Windows if the issue is high frequency simultaneous reading and writing over BT dongle/chipset in your PC. However, testing the gamepad without DS4Windows would almost like running the DS4Windows app with "Enable output data to DS4" option unticked (=DS4Windows only reads data from the gamepad, but doesn't try to send output data).

Anyway, could be that DS4Windows app is a bit too aggressive with "writeOutput failed" events. The app assumes the gamepad connection is disconnected right after the first error like this. In your PC things might work better if writeOutput process is done less frequently or DS4Windows would give up the connection only after like 10 repeated writeOutput errors (and not right after the first error case). I may tweak the debug version to support something like this, so you could test this theory.

```
08/11/2020 00:26:16: DEBUG: sendOutputReport. ERROR. writeOutput failed. Force quiting output thread. LastErrorCode=1167
```

**Comment by mika-n** [comment]
@Newtype14  The debug version has now a "delayed force quit" logic in BT read and write errors. The new debug version doesn't force disconnect the gamepad on the first read/write error. There needs to be handful of consecutive errors until DS4WindowsDebug app gives up and assumes the gamepad connection really is lost. Please test this version to see if it keeps the BT connection alive. If there are random read/write errors then then the Logs tabpage and logfile shows notification messages about those errors even when the gamepad is not disconnected from DS4Windows app until "too many" errors counted. 
https://github.com/Ryochan7/DS4Windows/wiki/Troubleshooting#detailed-debug-log-file-from-gamepad-connection-problems

**Comment by Newtype14** [comment]
@mika-n Just tested it. I didn't get disconnected. Controller didn't act weird or anything. There is something new in the log (warnings).
[Debug Log.txt](https://github.com/Ryochan7/DS4Windows/files/5080504/Debug.Log.txt)

**Comment by mika-n** [comment]
Those `DEBUG: performDs4Input. WARNING. 1C:96:5A:7B:86:5A crc32 check of BT input packet failed. btInputReport type0=17  type2=0`  warning messages are normal in certain scenarios and nothing to worry about (these warnings may be seen right after the game connection in BT if the gamepad or BT communication line is a bit slow to respond with certain DS4 messages. However, eventually the DS4 handshaking worked because only three lines of those warnings and no connection error msg).

The debug log file didn't show any of those "performDS4Output writeFailed" warnings. The debug version should show this message if writeOutput process fails, but does NOT disconnect the gamepad right away after the first writeOutput failure. However, writeOutput failure is shown in the logfile even when the error is "ignored and the gamepad usage is continue business as usual". Strange that the debug version doesn't suffer from the writeOutput failed case because it is identical and the same code as the original vanilla version of DS4Windows app.  If you switch over to use the original vanilla version of DS4Windows app then do you see those writeOutput failed/disconnection errors right away?

**Comment by Newtype14** [comment]
I guess I got really lucky the last time. I played two long sessions using the debug version and never got any errors. Then I played with the normal version and got an error and disconnect after an hour. Finally, I tried the debug version again and eventually got disconnected. Seems it just spams errors until the controller finally disconnects. Here's the log: 
[DS4Windows Debug Log 01.txt](https://github.com/Ryochan7/DS4Windows/files/5091184/DS4Windows.Debug.Log.01.txt)

I'm gonna try using the old XBOX 360 Windows 7 driver (xusb21) instead of the one that comes with Windows 10 (xusb22). It's the only thing I haven't tried at this point.

**Comment by Newtype14** [comment]
After a lot of testing, I found what the problem is. The controller is faulty. I tested another v1 controller for an entire week and never had any disconnects. It sucks to have wasted so much time trying to fix this problem only to come to this conclusion, but at least it's not a problem with DS4Windows. Thanks @mika-n for the help.

**Comment by mika-n** [comment]
OK. Good you found a "cure". DS4 gamepad has a factory reset feature to re-initialize the gamepad firmware settings. There is a small hole on the back of the gamepad you can press with a paper clip. Keep it down for 10 secs. Then re-pair the gamepad and see how it goes.

If you have a usb dongle then some gamepads have had a weak BT radio signal, but moving the dongle from the back of the PC to front of the PC may help to fight that issue. Or using the BT dongle at the end of USB extension cable (ie. taking the dongle further away from the radio storm caused by the PC chassis).

**Comment by Newtype14** [comment]
> There is a small hole on the back of the gamepad you can press with a paper clip. Keep it down for 10 secs. Then re-pair the gamepad and see how it goes.

I tried and it didn't work. I don't have a USB dongle, just a motherboard antenna, and it's about 80cm from my controller with nothing in-between.

I have two v1 controllers. The one that works was bought not long before the PS4 Pro came out. The one that keeps disconnecting was bought years before that. They both have the same model number, both are authentic, but I wonder if early models had BT issues.

**Comment by mika-n** [comment]
closed as "probably DS4 gamepad hardware issue".
